### PR TITLE
track non roslyn text buffer changes as well as roslyn text buffer chang...

### PR DESCRIPTION
**Pull request by heejaechang**
...es to delay solution crawler

There are 2 things incremental processor takes care of

@1 is making sure we delay processing any work until there is enough idle (ex, typing) in host.
@2 is managing cancellation and pending works.

we used to do #1 and #2 only for Roslyn files. and that is usually fine since most of time solution contains only roslyn files.

but for mixed solution (ex, Roslyn files + HTML + JS + CSS), #2 still makes sense but #1 doesn't. We want to pause any work while something is going on in other project types as well.

we need to make sure we play nice with neighbors as well.

now, we don't care where changes are coming from. if there is any change in host, we puase oursevles for a while.

### CODE REVIEW

**Comment by heejaechang** [comment]
@pharring @srivatsn @jasonmalinowski @rchande @shyamnamboodiripad can you take a look?

**Comment by heejaechang** [comment]
by the way, this is an attempt to reduce allocation in one of perf tests.

**Comment by jasonmalinowski** [comment]
Rather than couple all of this together, I'd feel better if there was a simple IWpfTextViewCreationListener that subscribed to events and then managed it's own state. That way it'd be a separate component entirely without shared state. I think it'll be less code, would have the advantage that it'll work in non-VS hosts, and doesn't churn what is one of the more subtle pieces of code in our codebase.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs [hunk 1]
```diff
@@ -300,35 +318,44 @@ private IList<DocumentId> GetDocumentIdsFromDocCookie(uint docCookie)
...
-                // context document between the closing of individual documents in the set. In the
-                // case of closing the tab of a shared document, this avoids updating the shared 
-                // item context hierarchy for the entire shared project to head project owning the
-                // last documentKey in this list.
-                var updateActiveContext = documentsToClose.Count == 1;
-
-                foreach (var documentKey in documentsToClose)
+                // let others know about non roslyn document close
+                var buffer = TryGetTextBufferFromDocData(RunningDocumentTable.GetDocumentData(docCookie));
+                if (buffer != null)
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs]: What happens if the document, after already being open, was then added to the workspace? At this point this code path will get skipped, and the view unsubscribe might not run?

Trying to do this state tracking in this class really isn't the right thing to do. We should just have a separate class to subscribe to the text view notifications.

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentTrackingService.cs [hunk 2]
```diff
@@ -262,5 +282,51 @@ internal string GetDebuggerDisplay()
...
+            {
+                _owner = owner;
+                _buffers = new HashSet<ITextBuffer>();
+            }
+
+            public void OnOpened(ITextBuffer buffer)
+            {
+                AssertIsForeground();
+
+                if (_buffers.Contains(buffer))
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentTrackingService.cs]: What happens for multiple views over the same buffer?

**Comment by heejaechang** [comment]
@jasonmalinowski @pharring @rchande can you take a look?

**Comment by heejaechang** [comment]
I talked to Jason already about using IWpfTextViewCreationListener. we can't use it due to several reasons. such as dll being loaded into VS even when Roslyn is not used or having circular reference unless we create new dll and etc.

File: src/VisualStudio/Core/Def/RoslynDocumentProvider.cs [hunk 3]
```diff
@@ -23,9 +26,50 @@ internal class RoslynDocumentProvider : DocumentProvider
...
+            if (view != null)
+            {
+                _documentTrackingService?.OnNonRoslynViewOpened(view);
+            }
+        }
+
+        private ITextView GetTextViewFromFrame(IVsWindowFrame frame)
+        {
+            IntPtr unknown;
+            if (ErrorHandler.Failed(frame.QueryViewInterface(typeof(IVsCodeWindow).GUID, out unknown)) || unknown == IntPtr.Zero)
```
Comment (pharring) [inline src/VisualStudio/Core/Def/RoslynDocumentProvider.cs]: This isn't the right way to get a code window from a frame. Microsoft.VisualStudio.Shell.VsShellUtilities has a GetTextView implementation which you should study. It deals with things such as split XAML/Designer views.

**Comment by pharring** [comment]
Please fix the GetTextView thing. Otherwise LGTM.

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentTrackingService.cs [hunk 4]
```diff
@@ -262,5 +276,52 @@ internal string GetDebuggerDisplay()
...
+            {
+                AssertIsForeground();
+
+                var view = sender as ITextView;
+                if (view == null || !_views.Contains(view))
+                {
+                    return;
+                }
+
+                view.TextBuffer.PostChanged -= OnTextChanged;
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentTrackingService.cs]: Don't forget to unsubscribe Closed as well.

**Comment by jasonmalinowski** [comment]
:+1: once Paul's comment is taken care of.

File: src/VisualStudio/Core/Def/RoslynDocumentProvider.cs (hunk 3 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/RoslynDocumentProvider.cs]: thank you

File: src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentTrackingService.cs (hunk 4 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioDocumentTrackingService.cs]: !!! thank you

**Comment by pharring** [comment]
@heejaechang Why not call the MPF helper in your implementation of GetTextView?

**Pull request merged**